src/persist-client/src/write.rs

```diff
@@ -341,44 +437,230 @@ where
+        updates: &[((K, V), T, D)],
+        expected_upper: T,
+        new_upper: T,
+    ) {
+        self.compare_and_append(
+            updates.iter().map(|((k, v), t, d)| ((k, v), t, d)),
+            Antichain::from_elem(expected_upper),
+            Antichain::from_elem(new_upper),
+        )
+        .await
+        .expect("external durability failed")
+        .expect("invalid usage")
+        .expect("unexpected upper")
+    }
+}
+
+/// A handle to a batch of updates that has been written to blob storage but
+/// which has not yet been appended to a shard.
+///
+/// A [Batch] needs to be marked as consumed or it needs to be deleted via [Self::delete].
+/// Otherwise, a dangling batch will leak and backing blobs will remain in blob storage.
+// TODO(aljoscha): Right now, we require batches to be manually marked as
+// consumed (when they're appended to a batch) or deleted. This is clunky and
+// prone to errors so we should replace it with lazy cleanup once we can use
+// async drop.
+#[derive(Debug)]
+pub struct Batch<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    /// [Description] of updates contained in this batch.
+    pub desc: Description<T>,
+
+    shard_id: ShardId,
+
+    /// Keys to blobs that make up this batch of updates.
+    blob_keys: Vec<String>,
+
+    /// Handle to the [BlobMulti] that the blobs of this batch were uploaded to.
+    blob: Arc<dyn BlobMulti + Send + Sync>,
+
+    // These provide a bit more safety against appending a batch with the wrong
+    // type to a shard.
+    _phantom: PhantomData<(K, V, T, D)>,
+}
+
+impl<K, V, T, D> Drop for Batch<K, V, T, D>
+where
+    T: Timestamp + Lattice + Codec64,
+{
+    fn drop(&mut self) {
+        if self.blob_keys.len() > 0 {
+            warn!(
+                "un-consumed Batch, with {} dangling blob keys: {:?}",
+                self.blob_keys.len(),
+                self.blob_keys
+            );
+        }
+    }
+}
+
+impl<K, V, T, D> Batch<K, V, T, D>
+where
+    K: Debug + Codec,
+    V: Debug + Codec,
+    T: Timestamp + Lattice + Codec64,
+    D: Semigroup + Codec64,
+{
+    pub(crate) fn new(
+        blob: Arc<dyn BlobMulti + Send + Sync>,
+        shard_id: ShardId,
+        desc: Description<T>,
+        blob_keys: Vec<String>,
+    ) -> Self {
+        Self {
+            desc,
+            blob_keys,
+            shard_id,
+            blob,
+            _phantom: PhantomData,
+        }
+    }
+
+    /// Marks the blobs that this batch handle points to as consumed, likely
+    /// because they were appended to a shard.
+    ///
+    /// Consumers of a blob need to make this explicit, so that we can log
+    /// warnings in case a batch is not used.
+    pub(crate) fn mark_consumed(&mut self) {
+        self.blob_keys.clear();
+    }
+
+    /// Deletes the blobs that make up this batch from the given blob store and
+    /// marks them as deleted.
+    pub async fn delete(&mut self) -> Result<(), ExternalError> {
```

Comment:
should this take `self`? I can't see why someone would want to reuse the batch after, can you? (the builder is eventually a different story, lots of nice allocs there to be potentially reused)

Reply:
The `_append()` methods take a `&mut Batch`, mostly so that callers can keep the batch and retry. And the cleanup-check logic becomes tricky if `Batch` is cloned. But I don't like that `delete` takes a `&mut self`, either. :see_no_evil:

Reply:
With this change: https://github.com/MaterializeInc/materialize/pull/12477#discussion_r879545137. It seems I can make `delete` take `self`, because all places that want to call it own the `Batch`.